### PR TITLE
feat: Add Substrait-based ODFV transformation

### LINKED ISSUE #3945: Substrait-based on demand feature views
**Is your feature request related to a problem? Please describe.**
On demand feature views as implemented right now are very limited. The only way to specify odfvs is through a python function that takes in pandas Dataframe as input and outputs another pandas Dataframe. This leads to problems for both offline and online interfaces:

- Even the most scalable offline stores are forced to collect the whole dataset as a single pandas Dataframe to apply odfv function. There's no way for offline stores to incorporate computation in their engines.
- udfs in odfvs are inherently bound to pandas and python runtime. Non-python feature servers are stuck with the problem of figuring out how to run this functions if necessary. Java feature server for example has a separate python transformation service only for this reason, but that's obviously a subpar solution as the whole point of a java feature server was to avoid python runtime in feature serving in the first place.

**Describe the solution you'd like**
Allow constructing odfvs as substrait plans. [Substrait](https://substrait.io/) is a protobuf-based serialization format for relational algebra operations. It is meant to be used as a cross-language and cross-engine format for sharing logical or physical execution plans. It has a number of producers (tools that can generate substrait) and consumers (engines that can run substrait) in different languages.

- Different offline stores will be able to inspect and incorporate substrait plans in their transformations. Even if that's impossible the default implementation inside feast to apply these functions will avoid pandas.
- Most importantly, non-python feature servers like a java feature server will be able to apply the functions without a separate python component. Apache Arrow java implementation comes with java bindings to Acero query engine that can consume substrait plans. (https://arrow.apache.org/docs/java/substrait.html#executing-queries-using-substrait-plans)

The example code in my PoC implementation looks something like this:

```
def generate_substrait():
    import ibis
    from ibis_substrait.compiler.core import SubstraitCompiler

    compiler = SubstraitCompiler()

    t = ibis.table([("conv_rate", "float"), ("acc_rate", "float")], "t")

    expr = t.select((t['conv_rate'] + t['acc_rate']).name('conv_rate_plus_acc_substrait'))

    return compiler.compile(expr).SerializeToString()

substrait_odfv = OnDemandFeatureView(
    name='substrait_view',
    sources=[driver_stats_fv],
    schema=[
        Field(name="conv_rate_plus_acc_substrait", dtype=Float64)
    ],
    substrait_plan=generate_substrait()
)
```
Substait plan object that feast accepts is `bytes` and introduces no external dependency. I'm using `ibis` and `ibis-substrait` to generate the plan. Right now that's the most practical way to generate substrait plan in python with DataFrame-like API, but this could have been any other substrait producer.

**Describe alternatives you've considered**
An obvious alternative to substrait is sql-based odfvs, but using SQL has a number of important downsides:

1. The presence of different sql dialects means that, it will be especially hard to ensure that sql-based feature functions will behave the same way across different offline store and online store implementations.
2. The user is implicitly bound to their offline store and online store of choice, because the dialect used in sql strings has to match offline store engine.

Having said that, it probably makes sense to support both substrait-based and sql-based odfvs, because at the moment it might be easier for sql-based logic to be incorporated inside offline store engines.

### PR DESCRIPTION
**Pull request by tokoko**
**What this PR does / why we need it**:
This PR introduces a new ODFV transformation type, based on [substrait](https://substrait.io/). The rationale behind the feature is described in an issue #3945. PR also adds a simple way for users to express substrait transformations using ibis. `ibis` and `ibis-substrait` librarie dependencies are added under an extra called `ibis` for users willing to use this feature. 

The way to express substrait-based ODFV is pretty similar to the existing pandas-based version, the only difference being that the user-defined function must have a return type of `ibis.expr.types.relations.Table` instead of `pandas.DataFrame`.

```
from ibis.expr.types import Table
@on_demand_feature_view(
    sources=[driver_stats_fv[["conv_rate", "acc_rate"]]],
    schema=[Field(name="conv_rate_plus_acc_substrait", dtype=Float64)],
)
def substrait_view(inputs: Table) -> Table:
    return inputs.select(
        (inputs["conv_rate"] + inputs["acc_rate"]).name("conv_rate_plus_acc_substrait")
    )
```

**Which issue(s) this PR fixes**:
Fixes #3945

### CODE REVIEW

File: sdk/python/tests/unit/test_on_demand_substrait_transformation.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+            df["conv_rate_plus_acc"] = inputs["conv_rate"] + inputs["acc_rate"]
+            return df
+
+        from ibis.expr.types import Table
+
+        @on_demand_feature_view(
+            sources=[driver_stats_fv[["conv_rate", "acc_rate"]]],
+            schema=[Field(name="conv_rate_plus_acc_substrait", dtype=Float64)],
+        )
+        def substrait_view(inputs: Table) -> Table:
```
Comment (HaoXuAI) [inline sdk/python/tests/unit/test_on_demand_substrait_transformation.py:75]: Something can try is creat a FeastTable or FeastDataframe class that wrap pandas df and other table etc.

Comment (tokoko) [reply sdk/python/tests/unit/test_on_demand_substrait_transformation.py:75]: You mean like an abstraction on top of both? But that would necessitate whole API for transformations around it and in the end we would essentially be recreating ibis.

Comment (HaoXuAI) [reply sdk/python/tests/unit/test_on_demand_substrait_transformation.py:75]: Maybe just a type class that can be accepting different apis. Anyway I'm not sure how to implement it yet. I think it looks good with your implementation

**Review by HaoXuAI — APPROVED** [review]

**Comment by HaoXuAI** [comment]
I think this is a great feature that should include in the documentation somewhere

**Comment by tokoko** [comment]
Thanks, I'll follow up shortly with a PR for doc changes.

**Comment by sudohainguyen** [comment]
hey this is a cool feature, would love to see the RFC or something similar to have a deeper understanding about your idea

**Comment by tokoko** [comment]
@sudohainguyen hey, I shared this [document](https://docs.google.com/document/d/1NEyp7J-cQTqFOX7haLHS7aa5VXapWnTwA_sb7dq1GOQ) a couple months back on slack. It's a first draft for a RFC, somewhat incomplete though. I should probably request access to RFC google folder and turn this into a real RFC document.

**Pull request merged**